datafusion/core/src/physical_optimizer/remove_unnecessary_sorts.rs

```diff
@@ -0,0 +1,887 @@
+                .map(transform)
+                .collect::<Result<Vec<_>>>()?;
+            let children_plans = children_requirements
+                .iter()
+                .map(|elem| elem.plan.clone())
+                .collect::<Vec<_>>();
+            let sort_onwards = children_requirements
+                .iter()
+                .map(|item| {
+                    if item.sort_onwards.is_empty() {
+                        vec![]
+                    } else {
+                        // TODO: When `maintains_input_order` returns Vec<bool>,
+                        //       pass the order-enforcing sort upwards.
+                        item.sort_onwards[0].clone()
+                    }
+                })
+                .collect::<Vec<_>>();
+            let plan = with_new_children_if_necessary(self.plan, children_plans)?;
+            Ok(PlanWithCorrespondingSort { plan, sort_onwards })
+        }
+    }
+}
+
+impl PhysicalOptimizerRule for RemoveUnnecessarySorts {
+    fn optimize(
+        &self,
+        plan: Arc<dyn ExecutionPlan>,
+        _config: &SessionConfig,
+    ) -> Result<Arc<dyn ExecutionPlan>> {
+        // Execute a post-order traversal to adjust input key ordering:
+        let plan_requirements = PlanWithCorrespondingSort::new(plan);
+        let adjusted = plan_requirements.transform_up(&remove_unnecessary_sorts)?;
+        Ok(adjusted.plan)
+    }
+
+    fn name(&self) -> &str {
+        "RemoveUnnecessarySorts"
+    }
+
+    fn schema_check(&self) -> bool {
+        true
+    }
+}
+
+fn remove_unnecessary_sorts(
+    requirements: PlanWithCorrespondingSort,
+) -> Result<Option<PlanWithCorrespondingSort>> {
+    // Perform naive analysis at the beginning -- remove already-satisfied sorts:
+    if let Some(result) = analyze_immediate_sort_removal(&requirements)? {
+        return Ok(Some(result));
+    }
+    let plan = &requirements.plan;
+    let mut new_children = plan.children().clone();
+    let mut new_onwards = requirements.sort_onwards.clone();
+    for (idx, (child, sort_onwards, required_ordering)) in izip!(
+        new_children.iter_mut(),
+        new_onwards.iter_mut(),
+        plan.required_input_ordering()
+    )
+    .enumerate()
+    {
+        let physical_ordering = child.output_ordering();
+        match (required_ordering, physical_ordering) {
+            (Some(required_ordering), Some(physical_ordering)) => {
+                let is_ordering_satisfied = ordering_satisfy_concrete(
+                    physical_ordering,
+                    required_ordering,
+                    || child.equivalence_properties(),
+                );
+                if !is_ordering_satisfied {
+                    // Make sure we preserve the ordering requirements:
+                    update_child_to_remove_unnecessary_sort(child, sort_onwards)?;
+                    let sort_expr = required_ordering.to_vec();
+                    *child = add_sort_above_child(child, sort_expr)?;
```

Comment:
It is somewhat strange to me that a pass called "RemoveUnecessarySorts" is actually sometimes adding sorts 🤔

Reply:
This code should work quite rarely. However, there are cases, where it is necessary. Assume we have a plan like following 
```sql
`Window Requires a DESC (Non reversible)`
  `Window Requires a DESC (Reversible)` 
    `Sort a DESC`
      `Window Requires a ASC`
        `Sort a ASC`
```
 during traversing the plan bottom up , we can remove second `Sort` since executor its above can calculate its result with reverse order. However, one SortExec may not always satisfies single executor. Hence removing a Sort may invalidate assumptions for executors above. This check handles these cases. Our rule turns above plan to the one below
```sql
`Window Requires a DESC (Non reversible)`
  `Sort a DESC`
    `Window Requires a DESC (Reversible)` 
      `Window Requires a ASC`
        `Sort a ASC`
``` 
without this addition we will generate plan below
``` sql
`Window Requires a DESC (Non reversible)`
  `Window Requires a DESC (Reversible)` 
    `Window Requires a ASC`
      `Sort a ASC`
``` 
which doesn't satisfy the requirement for the executor at the top.

Reply:
One thing we can do is to name the rule as ‘OptimizeSorts’ instead of ‘RemoveUnnecessarySorts’ so that others do not get thrown off by the sort adding code (which runs rarely, but needs to be there to cover corner cases @mustafasrepo explains).